File: src/main/java/com/rdvdev2/timetravelmod/impl/common/CustomTrigger.java
package com.rdvdev2.timetravelmod.impl.common;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import com.google.gson.JsonObject;
import net.minecraft.advancement.PlayerAdvancementTracker;
import net.minecraft.advancement.criterion.AbstractCriterion;
import net.minecraft.advancement.criterion.AbstractCriterionConditions;
import net.minecraft.advancement.criterion.Criterion;
import net.minecraft.predicate.entity.AdvancementEntityPredicateDeserializer;
import net.minecraft.predicate.entity.EntityPredicate;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Identifier;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class CustomTrigger extends AbstractCriterion<CustomTrigger.Instance> {

    private final Identifier identifier;
    public final Map<PlayerAdvancementTracker, Listeners> listeners = Maps.newHashMap();

    public CustomTrigger(Identifier identifier) {
        this.identifier = identifier;
    }

    @Override
    public Identifier getId() {
        return identifier;
    }

    @Override
    protected CustomTrigger.Instance conditionsFromJson(JsonObject obj, EntityPredicate.Extended playerPredicate, AdvancementEntityPredicateDeserializer predicateDeserializer) {
        return new Instance(getId(), playerPredicate);
    }

    public void trigger(ServerPlayerEntity player) {
        Listeners triggerListeners = listeners.get(player.getAdvancementTracker());

        if (triggerListeners != null) triggerListeners.trigger(player);
    }

    public static class Instance extends AbstractCriterionConditions {

        public Instance(Identifier identifier, EntityPredicate.Extended playerPredicate) {
            super(identifier, playerPredicate);
        }

        public boolean test() {
            return true;
        }
    }

    static class Listeners {
        private final PlayerAdvancementTracker playerAdvancements;
        private final Set<ConditionsContainer<?>> listeners = Sets.newHashSet();

        public Listeners(PlayerAdvancementTracker playerAdvancements) {
            this.playerAdvancements = playerAdvancements;
        }

        public boolean isEmpty() {
            return listeners.isEmpty();
        }

        public void add(Criterion.ConditionsContainer<?> listener) {
            listeners.add(listener);
        }

        public void remove(Criterion.ConditionsContainer<?> listener) {
            listeners.remove(listener);
        }

        public void trigger(ServerPlayerEntity player) {
            ArrayList<ConditionsContainer<?>> list = null;

            for (Criterion.ConditionsContainer<?> listener: listeners) {
                if ( ((Instance) listener.getConditions()).test() ) {
                    if (list == null) {
                        list = Lists.newArrayList();
                    }
                    list.add(listener);
                }
            }

            if (list != null) {
                for (Criterion.ConditionsContainer<?> listener: list)
                    listener.grant(playerAdvancements);
            }
        }
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/common/CustomTrigger.kt
package com.rdvdev2.timetravelmod.impl.common

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import com.google.common.collect.Sets
import com.google.gson.JsonObject
import net.minecraft.advancement.PlayerAdvancementTracker
import net.minecraft.advancement.criterion.AbstractCriterion
import net.minecraft.advancement.criterion.AbstractCriterionConditions
import net.minecraft.advancement.criterion.Criterion.ConditionsContainer
import net.minecraft.predicate.entity.AdvancementEntityPredicateDeserializer
import net.minecraft.predicate.entity.EntityPredicate.Extended
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier

class CustomTrigger(private val identifier: Identifier) : AbstractCriterion<CustomTrigger.Instance>() {

    private val listeners: Map<PlayerAdvancementTracker, Listeners> = Maps.newHashMap()

    override fun getId(): Identifier {
        return identifier
    }

    override fun conditionsFromJson(obj: JsonObject, playerPredicate: Extended, predicateDeserializer: AdvancementEntityPredicateDeserializer): Instance {
        return Instance(id, playerPredicate)
    }

    fun trigger(player: ServerPlayerEntity) {
        val triggerListeners = listeners[player.advancementTracker]
        triggerListeners?.trigger(player)
    }

    class Instance(identifier: Identifier?, playerPredicate: Extended?) : AbstractCriterionConditions(identifier, playerPredicate) {
        fun test(): Boolean {
            return true
        }
    }

    class Listeners(private val playerAdvancements: PlayerAdvancementTracker) {

        private val listeners: MutableSet<ConditionsContainer<*>> = Sets.newHashSet()

        val empty: Boolean
            get() = listeners.isEmpty()

        fun add(listener: ConditionsContainer<*>) {
            listeners.add(listener)
        }

        fun remove(listener: ConditionsContainer<*>) {
            listeners.remove(listener)
        }

        fun trigger(player: ServerPlayerEntity?) {
            var list: ArrayList<ConditionsContainer<*>>? = null
            for (listener in listeners) {
                if ((listener.conditions as Instance).test()) {
                    if (list == null) {
                        list = Lists.newArrayList()
                    }
                    list!!.add(listener)
                }
            }
            if (list != null) {
                for (listener in list) listener.grant(playerAdvancements)
            }
        }
    }
}